src/slice.ts

```diff
@@ -3,29 +3,39 @@ import createReducer from './reducer';
 import { createSelector, createSelectorName } from './selector';
 import { Action } from './types';
 
-type Reduce<State> = (state: State, payload?: any) => State | undefined | void;
+// type GetPayloadType<A> = A extends ((payload: infer P) => Action) ? P : A extends ((payload: infer P) => RAction) ? P : A;
+
+type Reduce<State, Payload = null> = Payload extends null
+  ? (state: State) => State | undefined | void
+  : (state: State, payload: Payload) => State | undefined | void;
+
 interface ReduceMap<State> {
-  [key: string]: Reduce<State>;
+  [key: string]: Reduce<State, any>;
 }
+
 interface ICreate<State, Actions> {
   slice?: string;
-  actions: { [key: string]: Reduce<State> };
+  actions: { [key in keyof Actions]: Reduce<State, Actions[key]> };
   initialState: State;
 }
 
+
+
 const actionTypeBuilder = (slice: string) => (action: string) =>
   slice ? `${slice}/${action}` : action;
 
+
 export default function create<
-  SliceState = any,
-  Actions = { [key: string]: (p?: any) => Action<any> },
+  SliceState,
+  Actions = undefined,
   State = any
->({ slice = '', actions, initialState }: ICreate<SliceState, Actions>) {
-  const actionKeys = Object.keys(actions) as (keyof Actions)[];
+  >({ slice = '', actions, initialState }: ICreate<SliceState, Actions extends undefined ? {[s:string]:any} :Actions >) {
```

Comment:
Yes, it's rather roundabout because typescripts type inference gets in the way if I use the Actions type directly. 
If used directly, typescript will ignore the fallback type in `Actions = {[key: string]: any}` and instead infer it from the first place it is used.
Hence the roundabout method.

Reply:
Yes!  I was running into this issue where the fallback seemed like it wasn't being used properly at all.  Nice!